services/src/main/java/org/keycloak/protocol/saml/SamlService.java

```diff
@@ -653,30 +654,39 @@ public String getDescriptor() throws Exception {
     }
 
     public static String getIDPMetadataDescriptor(UriInfo uriInfo, KeycloakSession session, RealmModel realm) {
-        Set<KeyWrapper> keys = new TreeSet<>((o1, o2) -> o1.getStatus() == o2.getStatus() // Status can be only PASSIVE OR ACTIVE, push PASSIVE to end of list
-          ? (int) (o2.getProviderPriority() - o1.getProviderPriority())
-          : (o1.getStatus() == KeyStatus.PASSIVE ? 1 : -1));
-        keys.addAll(session.keys().getKeys(realm, KeyUse.SIG, Algorithm.RS256));
-
         try {
-            List<Element> signingKeys = new ArrayList<Element>();
-            for (KeyWrapper key : keys) {
-                signingKeys.add(IDPMetadataDescriptor.buildKeyInfoElement(key.getKid(), PemUtils.encodeCertificate(key.getCertificate())));
-            }
+            List<Element> signingKeys = new LinkedList<>();
+            session.keys().getKeysStream(realm, KeyUse.SIG, Algorithm.RS256)
+                    .sorted(SamlService::compareKeys)
+                    .forEach(key -> {
+                        try {
+                            Element element = IDPMetadataDescriptor
+                                    .buildKeyInfoElement(key.getKid(), PemUtils.encodeCertificate(key.getCertificate()));
+                            signingKeys.add(element);
+                        } catch (ParserConfigurationException e) {
+                            throw new RuntimeException(e);
+                        }
+                    });
```

Comment:
Wouldn't it be better to map the keys to elements and then collect into a list at the end instead of populating the list as part of a forEach?

Reply:
@martin-kanis Overall this really looks good to me, thanks for the great work. Just one minor suggestion as you can see above.

Reply:
Right, changed to map + collect. Thank you Stefan for the review.